libraries/chain/controller.cpp

```diff
@@ -4478,18 +4479,20 @@ struct controller_impl {
                // Remove the block that threw and all forks built off it.
                fork_db.remove( (*ritr)->id() );
 
-               // pop all blocks from the bad fork, discarding their transactions
-               // ritr base is a forward itr to the last block successfully applied
-               auto applied_itr = ritr.base();
-               for( auto itr = applied_itr; itr != new_head_branch.end(); ++itr ) {
-                  pop_block();
-               }
-               EOS_ASSERT( !switch_fork || chain_head.id() == old_head_branch.back()->header.previous, fork_database_exception,
-                           "loss of sync between fork_db and chainbase during fork switch reversal" ); // _should_ never fail
+               if (switch_fork) {
```

Comment:
Why this `if (switch_fork)`? The comment at line 4479 makes sense to me that if a block throws, we should not keep its descendants.